core/src/main/java/org/mule/runtime/core/internal/config/CustomService.java

```diff
@@ -54,7 +68,61 @@ public Optional<Class> getServiceClass() {
    * @return the service implementation.
    */
   public Optional<Object> getServiceImpl() {
-    return serviceImpl;
+    return getServiceImpl(null);
+  }
+
+  public Optional<Object> getServiceImpl(Object defaultService) {
+    if (!serviceImplInterceptorConsumer.isPresent()) {
+      return empty();
+    }
+
+    DefaultServiceInterceptor serviceInterceptor = new DefaultServiceInterceptor(defaultService);
+    serviceImplInterceptorConsumer.get().accept(serviceInterceptor);
+
+    return serviceInterceptor.isRemove() ? empty() : ofNullable(serviceInterceptor.getNewServiceImpl());
+  }
+
+  private static class DefaultServiceInterceptor implements ServiceInterceptor {
+
+    private final Object serviceImpl;
+    private Object newServiceImpl;
+    private boolean remove;
+
+    public DefaultServiceInterceptor(Object serviceImpl) {
+      this.serviceImpl = serviceImpl;
+    }
+
+    @Override
+    public Optional<Object> getDefaultServiceImpl() {
+      return ofNullable(serviceImpl);
+    }
+
+    @Override
+    public void newServiceImpl(Object newServiceImpl) {
+      this.newServiceImpl = newServiceImpl;
+    }
+
+    @Override
+    public void skip() {
+      if (newServiceImpl != null) {
+        throw new IllegalStateException("A 'newServiceImpl' is already present");
+      }
+
+      remove = true;
+    }
+
+    public Object getNewServiceImpl() {
+      if (remove) {
+        throw new IllegalStateException("Service set to be removed, can't be overridden");
```

Comment:
improve the error message, give context tothe developer of what went wrong (add object key, object string representation, ...)